homeassistant/components/switchbot/config_flow.py

```diff
@@ -59,62 +70,128 @@ async def async_step_bluetooth(
-                ):
-                    continue
-                parsed = parse_advertisement_data(
-                    discovery_info.device, discovery_info.advertisement
-                )
-                if parsed and parsed.data.get("modelName") in SUPPORTED_MODEL_TYPES:
-                    self._discovered_advs[address] = parsed
+            if parsed and parsed.data.get("modelName") in SUPPORTED_MODEL_TYPES:
+                self._discovered_advs[address] = parsed
 
         if not self._discovered_advs:
-            return self.async_abort(reason="no_unconfigured_devices")
-
-        data_schema = vol.Schema(
-            {
-                vol.Required(CONF_ADDRESS): vol.In(
-                    {
-                        address: f"{parsed.data['modelName']} ({address})"
-                        for address, parsed in self._discovered_advs.items()
-                    }
-                ),
-                vol.Required(CONF_NAME): str,
-                vol.Optional(CONF_PASSWORD): str,
-            }
+            raise AbortFlow("no_unconfigured_devices")
+
+    async def _async_set_device(self, discovery: SwitchBotAdvertisement) -> None:
+        """Set the device to work with."""
+        self._discovered_adv = discovery
+        address = discovery.address
+        await self.async_set_unique_id(
+            format_unique_id(address), raise_on_progress=False
         )
+        self._abort_if_unique_id_configured()
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle the user step to pick discovered device."""
+        errors: dict[str, str] = {}
+        device_adv: SwitchBotAdvertisement | None = None
+        if user_input is not None:
+            device_adv = self._discovered_advs[user_input[CONF_ADDRESS]]
+            await self._async_set_device(device_adv)
+            if device_adv.data["isEncrypted"]:
+                return await self.async_step_password()
+            return await self._async_create_entry_from_discovery(user_input)
+
+        self._async_discover_devices()
+        if len(self._discovered_advs) == 1:
+            # If there is only one device we can ask for a password
+            # or simply confirm it
+            device_adv = list(self._discovered_advs.values())[0]
+            await self._async_set_device(device_adv)
+            if device_adv.data["isEncrypted"]:
+                return await self.async_step_password()
+            return await self.async_step_confirm()
+
         return self.async_show_form(
-            step_id="user", data_schema=data_schema, errors=errors
+            step_id="user",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(CONF_ADDRESS): vol.In(
+                        {
+                            address: name_from_discovery(parsed)
+                            for address, parsed in self._discovered_advs.items()
```

Comment:
can we filter out the ones that are already set up?

Reply:
hm that might be confusing for the user as they will expect to see all devices. Would be nice to have "disabled" entries.

Reply:
> can we filter out the ones that are already set up? 

👍
It will (see line 137).

Reply:
Verified there was a test for this as well `test_user_setup_wohand_already_configured`